tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/aux/ProtoSeqOps.scala

```diff
@@ -0,0 +1,268 @@
+package at.forsyte.apalache.tla.bmcmt.rules.aux
+
+import at.forsyte.apalache.tla.bmcmt.types.{CellT, UnknownT}
+import at.forsyte.apalache.tla.bmcmt.{Arena, ArenaCell, SymbState, SymbStateRewriter}
+import at.forsyte.apalache.tla.lir.TypedPredefs._
+import at.forsyte.apalache.tla.lir.convenience.tla
+import at.forsyte.apalache.tla.lir._
+
+/**
+ * Proto sequences that contain the absolute minimum for implementing TLA+ sequences. A proto sequence is essentially an
+ * array of fixed size (which we call 'capacity'). A sequence is a pair that consists of the length (an integer cell)
```

Comment:
The situation here is similar to vectors in C++. We allocate a fixed number of cells (capacity) that initially accommodate exactly the number of the elements that are required to construct a sequence (length). The actual length of the the sequence is not always known. For instance, `SubSeq(<<1, 2, 3>>, i, j)` would create a sequence of length 2 when `i = 2` and `j = 3`, but the proto sequence would have capacity of 3, because we cannot statically compute the length. The only thing we know is that the capacity is not increasing.